### PR TITLE
Refactor mapping into consistent and conservative mapping functions

### PR DESCRIPTION
**Pull request by DavidSCN**
## Main changes of this PR

Splits the `map` function into `mapConsistent` and `mapConservative` and implements the `hasComputedMapping` in the base class (instead of each derived class).

## Motivation and additional information

The mapping functions follow the same structure namely

```cpp
PRECICE_ASSERT(...)
if (conservative)
{...}
else
{...}
if(scaledConsistent)
 applyScaledConsistent
```
The map function itself becomes messy quite fast. The new interface defines the structure in the base class and requires the individual branches to be implemented in the derived classes. We do this already in the RBF mapping classes. Open for discussions.

## Author's checklist

* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.16.3.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

**Review by fsimonis — APPROVED** [review]
Thanks for tidying this up!
Looks all sane.

File: src/mapping/Mapping.cpp [hunk 1]
```diff
@@ -87,6 +87,31 @@ bool Mapping::requireGradient() const
...
+    mapConservative(inputDataID, outputDataID);
+  } else {
+    mapConsistent(inputDataID, outputDataID);
+  }
+  if (hasConstraint(Mapping::SCALEDCONSISTENT)) {
+    scaleConsistentMapping(inputDataID, outputDataID);
+  }
+}
+
 void Mapping::scaleConsistentMapping(int inputDataID, int outputDataID) const
```
Comment (boris-martin) [inline src/mapping/Mapping.cpp:115]: Slightly tangential to this PR, but couldn't this function be refactored to split the validity check and the actual scaling ?

File: src/mapping/Mapping.cpp [hunk 2]
```diff
@@ -87,6 +87,31 @@ bool Mapping::requireGradient() const
...
+  PRECICE_ASSERT(output()->data(outputDataID)->values().size() / output()->data(outputDataID)->getDimensions() == static_cast<int>(output()->vertices().size()),
+                 output()->data(outputDataID)->values().size(), output()->data(outputDataID)->getDimensions(), output()->vertices().size());
+
+  if (hasConstraint(CONSERVATIVE)) {
+    mapConservative(inputDataID, outputDataID);
+  } else {
+    mapConsistent(inputDataID, outputDataID);
+  }
+  if (hasConstraint(Mapping::SCALEDCONSISTENT)) {
+    scaleConsistentMapping(inputDataID, outputDataID);
```
Comment (boris-martin) [inline src/mapping/Mapping.cpp:111]: I'm not a big fan of this structure, with an if coming after an else. Wouldn't it be cleaner to have something like 

```cpp
if (hasConstraint(CONSERVATIVE)) {
mapConservative(...)
} else if (hasConstraint(CONSISTENT) {
mapConsisten(...);
} else if {hasConstraint(SCALEDCONSISTENT) {
mapConsistent(...);
scaleConsistentMapping(...);
}
```

Also, we have `CONSERVATIVE` but `Mapping::SCALEDCONSISTENT`. Prefix isn't consistent.

**Review by boris-martin — APPROVED** [review]
Great refactoring. I added some comments on what could be future improvements, but the current changes seem very good.

File: src/mapping/Mapping.cpp (hunk 1 above)
Comment (DavidSCN) [reply src/mapping/Mapping.cpp:115]: You mean the `scaledConsistentMapping` function? Yes, sure. I won't touch this part in this PR, though.

File: src/mapping/Mapping.cpp (hunk 2 above)
Comment (DavidSCN) [reply src/mapping/Mapping.cpp]: Yes, I adapted it to the structure you proposed here.

Comment (DavidSCN) [reply src/mapping/Mapping.cpp]: ... and fixed the inconsistency.

**Pull request merged**